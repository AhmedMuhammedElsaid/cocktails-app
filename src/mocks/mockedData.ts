export const cocktailDetailsMockedData = {
  idDrink: "14608",
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strAlcoholic: "Alcoholic",
  strGlass: "Highball glass",
  strInstructions:
    "Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale.",
  strInstructionsDE:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
  strInstructionsIT: `Versare 30gr di rum speziato in un bicchiere highball con ghiaccio.Riempi con Ginger Ale.`,
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strIngredient1: `Spiced rum`,
  strIngredient2: `Ginger ale`,
  strCreativeCommonsConfirmed: `No`,
  dateModified: `2016-11-15 11:29:27`,
};
export const cocktailListMockedData = [
  cocktailDetailsMockedData,
  { ...cocktailDetailsMockedData, idDrink: "14709" },
];
export const cocktailCardMockedData = {
  strDrink: "Cream Soda",
  strCategory: "Cocktail",
  strDrinkThumb: `https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg`,
  strInstructions:
    "Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .",
};
