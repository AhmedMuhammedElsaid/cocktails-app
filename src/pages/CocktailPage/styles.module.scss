@import "../../scss//variables.scss";
.HomeBtn{
  margin-block-start: 2rem;
    padding:.5rem 1rem;
    display: inline-block;
    text-transform: uppercase;  
    color:$mainWhite;
    background:$primaryColor;
    text-decoration: none;
    letter-spacing: .3rem;
    outline: 2px solid $primaryColor;
    outline-offset:4px ;
    border: none;
    transition: $mainTransition;
  
  &:hover{
    background: $darkGrey;
    text-decoration: none;
    color: $primaryColor;
    outline-color: $darkGrey;
  }
}
.DetailsPage {
  &__Container {
    display: flex;
    justify-content: center;
    align-items: center;
    margin: 0 auto;
    margin-block-start: 5rem;
    &__Img {
      border-radius: 1rem;
      width: 100%;
      height: 100%;
    }
    &__Details {
      h3 {
        font-weight: 700;
        letter-spacing: 0.3rem;
        letter-spacing: 0.3rem;
        text-transform: capitalize;
        margin-block-end: 0.5em;
      }
      h5 {
        margin-block-end: 0.5em;
      }
    }
  }
}
:global{
  .react-reveal{
    width: 100% ;
  }
}
