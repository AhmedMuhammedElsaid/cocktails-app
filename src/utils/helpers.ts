/*
 @params object
 return object with no null ,undefined , empty string "cleaned object"
*/
export function cleanObject(object: { [x: string]: any }) {
  if (!object) return {};
  if (typeof object === "object") {
    for (let propName in object) {
      if (
        object[propName] === null ||
        object[propName] === undefined ||
        object[propName]?.length === 0
      ) {
        delete object[propName];
      }
    }
  }
  return object;
}
