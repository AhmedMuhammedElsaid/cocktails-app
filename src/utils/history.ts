import { createBrowserHistory } from "history";

const history = createBrowserHistory();
// method to redirect the user to error page in case any error happens while fetching the data from the server
export const redirectToErrorPage = () => {
  history.push(`/pagenotfound`);
  history.go(0);
};
export default history;
