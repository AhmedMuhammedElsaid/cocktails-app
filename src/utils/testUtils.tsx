import { render, RenderResult } from "@testing-library/react";
import { BrowserRouter as Router } from "react-router-dom";

// this is a handy function that I would utilize for any component
// that relies on the router being in context
export const customRender = (component: JSX.Element): RenderResult => {
  return {
    ...render(<Router>{component}</Router>),
  };
};
