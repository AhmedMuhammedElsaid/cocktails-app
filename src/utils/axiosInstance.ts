import axios from "axios";
import { redirectToErrorPage } from './history';
import { toast } from "react-toastify";

const axiosInstance = axios.create({
  // setting the default uri for this instance
  baseURL: process.env.REACT_APP_BASE_URI,
});

// setting language  configuration on every single request we send to the BE
axiosInstance.interceptors.request.use(
  (request: any) => {
    return {
      ...request,
      headers: {
        ...request.headers,
        // setting the header lang german "de"
        "Accept-Language": "de",
      },
    };
  },
  (error) => {
    return Promise.reject(error);
  }
);

axiosInstance.interceptors.response.use(
  (response) => {
    return response.data;
  },
  (error) => {
    switch (error?.response?.status) {
      case 400:
        return {
          data: null,
          hasError: true,
          error: [error.response.data],
        };
      case 401:
        toast.error("Unauthorized Call Access");
        break;
      case 404:
        toast.error("404 Not Found");
        break;
      case 500:
        toast.error("Server Is Down");
        break;
    }
    redirectToErrorPage()
    return Promise.reject(error);
  }
);
export default axiosInstance;
