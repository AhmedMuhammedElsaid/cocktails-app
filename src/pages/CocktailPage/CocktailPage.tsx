import React, { FunctionComponent, useEffect, useRef, useState } from "react";
import { useParams } from "react-router";
import { CardMedia, Divider, Grid, Typography } from "@material-ui/core";
import Loader from "components/Loader";
import axiosInstance from "utils/axiosInstance";
import { Fade } from "react-reveal";
import { AxiosResponse, cocktailSchema } from "utils/types";
import styles from "./styles.module.scss";
import { cleanObject } from "utils/helpers";
import CocktailDetails from "../../components/CocktailDetails";
import { redirectToErrorPage } from "utils/history";
import { Link } from "react-router-dom";
import CocktailCard from "components/CocktailCard";
import CocktailModal from "../../components/CocktailModal";

interface Props {
  cocktail?: cocktailSchema;
}
const CocktailPage: FunctionComponent<Props> = ({ cocktail }) => {
  // grab the id to the selected cocktail.
  const { id }: { id: string } = useParams();

  const [currentCocktail, setCurrentCocktail] = useState<cocktailSchema | null>(
    cocktail || null
  );
  const ingredientNameRef = useRef("");
  const [isOpen, setIsOpen] = useState(false);

  // close & open modal methods
  const handleOpenModal = () => setIsOpen(true);
  const handleCloseModal = () => setIsOpen(false);
  // fetch the selected cocktail and render it's details
  const handleGetCocktailDetails = async () => {
    const { drinks }: AxiosResponse = await axiosInstance.get(
      `lookup.php?i=${id}`
    );
    if (drinks) {
      // remove null values from the cocktail as it being sent with many null values
      const cleanedCocktail: cocktailSchema = cleanObject({ ...drinks[0] });
      setCurrentCocktail(cleanedCocktail);
    } else {
      redirectToErrorPage();
    }
  };

  const [similarCocktail, setSimilarCocktail] = useState<
    cocktailSchema[] | null
  >(null);

  // fetch the similar cocktails which use the same ingredient
  const handleGetSimilarCocktails = async (ingredient: string) => {
    const { drinks }: AxiosResponse = await axiosInstance.get(
      `filter.php?i=${ingredient}`
    );
    setSimilarCocktail(drinks);
  };
  const handleGetSimilarCocktailToIngredient = async (
    ingredientName: string
  ) => {
    //performance balance, this condition cash the data for same ingredient ,
    // Ex, user click over coffee , modal will show up with cocktails use coffee , if the user close the modal and click coffee again,
    // i wont fire an api call since i already have the vodica cocktails cashed
    if (ingredientName !== ingredientNameRef.current) {
      // get similar cocktail that use the same ingredient
      await handleGetSimilarCocktails(ingredientName);
    }
    //store ingredient name to render it
    ingredientNameRef.current = ingredientName;

    handleOpenModal();
  };
  useEffect(() => {
    handleGetCocktailDetails();
  }, [id]);

  if (!currentCocktail) return <Loader />;
  return (
    <>
      <Link className={styles.HomeBtn} to="/" data-testid="home-btn-testid">
        Home
      </Link>
      <Grid container className={styles.DetailsPage}>
        <Fade bottom duration={2000} distance="40px">
          <Grid
            container
            item
            sm={10}
            className={styles.DetailsPage__Container}
            spacing={4}
          >
            <Grid item xs={12} sm={6} md={6} lg={6}>
              <CardMedia
                className={styles.DetailsPage__Container__Img}
                component="img"
                title="Cocktail Image"
                height="250"
                data-testid="cocktail-img-testid"
                image={currentCocktail?.strDrinkThumb}
                alt="cocktail pic"
              />
            </Grid>
            <Grid item xs={12} sm={6} md={6} lg={6}>
              <Typography variant="h4">{currentCocktail?.strDrink}</Typography>
              <Typography variant="h5">
                {currentCocktail?.strCategory}
              </Typography>
              {
                <CocktailDetails
                  currentCocktail={currentCocktail}
                  handleGetSimilarCocktailToIngredient={
                    handleGetSimilarCocktailToIngredient
                  }
                />
              }
            </Grid>
          </Grid>
        </Fade>
      </Grid>
      {/* Modal Content that popups while clicking over Ingredient with similar cocktails that use the same ingredient which clicked */}
      <CocktailModal onClose={handleCloseModal} open={isOpen}>
        <Fade top duration={1000}>
          <h2>{`
              cocktails that also use ${ingredientNameRef.current}
            `}</h2>
          <h3>Click Over The Cocktail To View It's Details</h3>
        </Fade>
        <Divider light style={{ margin: "1.5rem 0" }} />
        <Fade bottom duration={1000}>
          <Grid container>
            {similarCocktail ? (
              <Grid container spacing={2} onClick={handleCloseModal}>
                {similarCocktail?.map((cocktail, index) => (
                  <CocktailCard hideDesc cocktail={cocktail} key={index} />
                ))}
              </Grid>
            ) : (
              <Fade top duration={1000}>
                <Loader />
              </Fade>
            )}
          </Grid>
        </Fade>
      </CocktailModal>
    </>
  );
};

export default CocktailPage;
