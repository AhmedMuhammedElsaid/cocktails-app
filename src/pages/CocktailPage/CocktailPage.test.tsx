import CocktailPage from "./CocktailPage";
import { customRender } from "utils/testUtils";
import "@testing-library/jest-dom";
import { cocktailDetailsMockedData } from "mocks/mockedData";

const cocktailImgHolder = "cocktail-img-testid",
  cocktailsDetailsHolder = "cocktail-details-testid",
  homeBtn = "home-btn-testid";
describe("Test Cocktail Details Page", () => {
  it("should render the cocktail details page without crashing", () => {
    const { asFragment, getByTestId } = customRender(
      <CocktailPage cocktail={cocktailDetailsMockedData} />
    );
    expect(getByTestId(homeBtn)).toBeInTheDocument();
    expect(getByTestId(cocktailImgHolder)).toBeInTheDocument();
    expect(getByTestId(cocktailsDetailsHolder)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
