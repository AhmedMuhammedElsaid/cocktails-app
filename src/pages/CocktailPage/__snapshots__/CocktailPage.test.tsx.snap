// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Cocktail Details Page should render the cocktail details page without crashing 1`] = `
<DocumentFragment>
  <a
    class="HomeBtn"
    data-testid="home-btn-testid"
    href="/"
  >
    Home
  </a>
  <div
    class="MuiGrid-root DetailsPage MuiGrid-container"
  >
    <div
      class="react-reveal"
    >
      <div
        class="MuiGrid-root DetailsPage__Container MuiGrid-container MuiGrid-spacing-xs-4 MuiGrid-item MuiGrid-grid-sm-10"
      >
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <img
            alt="cocktail pic"
            class="MuiCardMedia-root DetailsPage__Container__Img MuiCardMedia-media MuiCardMedia-img"
            data-testid="cocktail-img-testid"
            height="250"
            src="https://www.thecocktaildb.com/images/media/drink/yqstxr1479209367.jpg"
            title="Cocktail Image"
          />
        </div>
        <div
          class="MuiGrid-root MuiGrid-item MuiGrid-grid-xs-12 MuiGrid-grid-sm-6 MuiGrid-grid-md-6 MuiGrid-grid-lg-6"
        >
          <h4
            class="MuiTypography-root MuiTypography-h4"
          >
            Cream Soda
          </h4>
          <h5
            class="MuiTypography-root MuiTypography-h5"
          >
            Cocktail
          </h5>
          <div
            data-testid="cocktail-details-testid"
          >
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                idDrink : 
              </span>
              14608
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                Alcoholic : 
              </span>
              Alcoholic
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                Glass : 
              </span>
              Highball glass
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                Instructions : 
              </span>
              Pour 1oz of Spiced Rum into a highball glass with ice. Fill with Ginger Ale.
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                InstructionsDE : 
              </span>
              Geben Sie 3 cl. Spiced Rum in ein Highball-Glas mit Eis. Mit Ginger Ale .
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                InstructionsIT : 
              </span>
              Versare 30gr di rum speziato in un bicchiere highball con ghiaccio.Riempi con Ginger Ale.
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Ingredient MuiTypography-h6"
            >
              Ingredient1 : Spiced rum
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Ingredient MuiTypography-h6"
            >
              Ingredient2 : Ginger ale
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                CreativeCommonsConfirmed : 
              </span>
              No
            </h6>
            <h6
              class="MuiTypography-root CocktailDetails__Info MuiTypography-h6"
            >
              <span>
                dateModified : 
              </span>
              2016-11-15 11:29:27
            </h6>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
